tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/client/MessageHandlerTest.java

```diff
@@ -117,6 +118,42 @@ public void onMessage(final ClientMessage message) {
       session.close();
    }
 
+   @Test
+   public void testMessageHandlerCloseTimeout() throws Exception {
+      // create Netty acceptor so client can use new onMessageCloseTimeout URL parameter
+      server.getRemotingService().createAcceptor("netty", "tcp://127.0.0.1:61616").start();
+      final int timeout = 1000;
+      locator = ActiveMQClient.createServerLocator("tcp://127.0.0.1:61616?onMessageCloseTimeout=" + timeout);
+      sf = createSessionFactory(locator);
+      ClientSession session = addClientSession(sf.createSession(false, true, true));
+      session.createQueue(QueueConfiguration.of(QUEUE).setDurable(false));
+      ClientProducer producer = session.createProducer(QUEUE);
+      producer.send(createTextMessage(session, "m"));
+
+      ClientConsumer consumer = session.createConsumer(QUEUE, null, false);
+
+      session.start();
+
+      CountDownLatch latch = new CountDownLatch(1);
+      consumer.setMessageHandler(message -> {
+         latch.countDown();
+         // don't just Thread.sleep() here because it will be interrupted on ClientConsumer.close()
+         long start = System.currentTimeMillis();
+         while (System.currentTimeMillis() - start < 2000) {
+            try {
+               Thread.sleep(100);
+            } catch (InterruptedException e) {
+               // ignore
+            }
+         }
+      });
+      latch.await();
+      long start = System.currentTimeMillis();
+      consumer.close();
+      long end = System.currentTimeMillis();
+      assertTrue( (end - start >= timeout) && (end - start <= 2000), "Closing consumer took " + (end - start) + "ms");
```

Comment:
I'd be inclined to use a shorter timeout for the close as Dom suggests, thus reducing the test duration, but also use a longer/indefinite 'dont complete yet' guard in the handler to be sure it is always still running after/when the timeout expires and can then 'immediately complete' on request, e.g instead arrange its waiting/exit-gating on an [atomic]boolean and other latch to wait on the rest of the test then complete quickly on request rather than the current time loop.

Reply:
Done.

Reply:
This actually seems more brittle now than it was when I commented on it originally.

Faster than before, yes (though could still be faster, if not just relying on friendly scheduling), but also more brittle as a result of now needing separately scheduled things to always happen <600ms apart, rather than <1000ms before.

Governing the exit point of the handler rather than just the time would entirely avoid the brittle reliance on nice scheduling.

Reply:
> Governing the exit point of the handler rather than just the time would entirely avoid the brittle reliance on nice scheduling.

I'm not following here. Can you elaborate?

Reply:
As far as I can tell the test is essentially relying on `java.util.concurrent.CountDownLatch#await(long, java.util.concurrent.TimeUnit)` (eventually called from `org.apache.activemq.artemis.api.core.client.ClientConsumer#close`) for its accuracy/reliability. Is that really reliant on nice scheduling?

Reply:
I push a change. Let me know what you think.

Reply:
You stopped the thread hanging around much longer than the checks, which is good...however the checks still remain as brittle to scheduling as they were before the change.

The main point is surely that it did not exit during the close timeout, and that the close takes at least the timeout. So you control the handler thread cant exit while those checks are happening, but then exits immediately as soon as they have completed, once you allow it to. The 800ms magic window doesnt need to be there.

Is it likely to take that long? No. Though could it happen now and then, on e.g a slow CI env with low thread count and shared resources? Yes, when someone then needs to go look at the sporadic failure. Or it adds to the many that have sporadically failed and encouraged people not to look.

If you want to use a timeout in the handler that fair enough, but I'd make it a lot longer so it can never reasonably happen unless the test has hung elsewhere.

It also needs add a try-finally to ensure the gating boolean is tripped to have it exit immediately even if the asserts somehow failed.

Reply:
I think I've got it now.

Reply:
Thats better, though caused me to realise something about it not guarding agains unexpected-exits. I added a new comment on that and readability, so I'll resolve this one.